homeassistant/components/thermostat/nest.py

```diff
@@ -97,9 +106,16 @@ def is_away_mode_on(self):
         """ Returns if away mode is on. """
         return self.structure.away
 
+    def enforce_temp_units(self, temperature):
+        """ Enforces temp units in C for nest API, returns temp """
+        if self.hass.config.temperature_unit == self.unit_of_measurement:
+            return temperature
+        else:
+            return self.nest_utils.f_to_c(temperature)
+
     def set_temperature(self, temperature):
         """ Set new target temperature """
-        self.device.target = temperature
+        self.device.target = self.enforce_temp_units(temperature)
```

Comment:
This should be solved in the component instead of the platform: we should only call `set_temperature` with the unit as specified in `unit_of_measurement`. I have extracted the helper methods that Home Assistant used internally so you can use them. They are available in the latest dev branch in the file [`homeassistant.helpers.temperature`](https://github.com/balloob/home-assistant/blob/086961d1093b466583c9c8125515981b242c0a2a/homeassistant/helpers/temperature.py) and [`homeassistant.util.temperature`](https://github.com/balloob/home-assistant/blob/086961d1093b466583c9c8125515981b242c0a2a/homeassistant/util/temperature.py)